File: 2021/10/09/lemonade_change.py
"""
https://leetcode.com/problems/lemonade-change/
Author: Andrew Jarombek
Date: 10/9/2021
"""

from typing import List


def lemonade_change(bills: List[int]) -> bool:
    fives = 0
    tens = 0

    for bill in bills:
        if bill == 5:
            fives += 1
        elif bill == 10:
            if fives > 0:
                fives -= 1
            else:
                return False

            tens += 1
        else:
            if tens > 0 and fives > 0:
                tens -= 1
                fives -= 1
            elif fives > 2:
                fives -= 3
            else:
                return False

    return True

File: 2022/05/06/lemonadeChange.ts
/**
 * https://leetcode.com/problems/lemonade-change/
 * @author Andrew Jarombek
 * @since 5/6/2022
 */

interface Change {
    5: number;
    10: number;
    20: number;
}

function lemonadeChange(bills: number[]): boolean {
    let change: Change = {
        5: 0,
        10: 0,
        20: 0
    }

    for (let bill of bills) {
        change[bill] = change[bill] + 1;
        bill -= 5;

        while (bill >= 20 && change[20] > 0) {
            bill -= 20;
            change[20] = change[20] - 1;
        }

        while (bill >= 10 && change[10] > 0) {
            bill -= 10;
            change[10] = change[10] - 1;
        }

        while (bill >= 5 && change[5] > 0) {
            bill -= 5;
            change[5] = change[5] - 1;
        }

        if (bill > 0) {
            return false;
        }
    }

    return true;
}